Document Ether.js for asset hub smart contracts

nhussein11 opened the pull request:
This PR aims to add a basic documentation of ether.js to showcase the interaction with asset hub smart contracts

_Note:_ **The IA used is preliminary**. We can circle back on this later on and define where all this content should be located

CrackTheCode016 commented on develop/toolkit/api-libraries/smart-contracts/ether-js.md:
```diff
@@ -0,0 +1,111 @@
+
+## Compile Contracts
+
+To deploy smart contracts to Asset Hub, you need to compile them into `polkavm` bytecode. Use the [`@parity/revive`](https://www.npmjs.com/package/@parity/revive){target=\_blank} library, which compiles Solidity code for use on Substrate-based chains.
```
Might be worth linking something about what PolkaVM is eventually, not sure if we have a good page here, if not we can make a follow up issue as I don't think it should block this PR.

CrackTheCode016 commented on develop/toolkit/api-libraries/smart-contracts/ether-js.md:
````diff
@@ -0,0 +1,111 @@
+```
+### Example: Storage.sol
+
+Here's a sample Solidity contract (`Storage.sol`) to be compiled and deployed to Asset Hub. This contract's functionality stores a number and permits users to update it with a new value.
````
tiny nit: not sure if we are saying "the Asset Hub" or just "Asset Hub" consistently.

```suggestion
Here's a sample Solidity contract (`Storage.sol`) to be compiled and deployed to the Asset Hub. This contract's functionality stores a number and permits users to update it with a new value.
```

CrackTheCode016 commented on develop/toolkit/api-libraries/smart-contracts/ether-js.md:
```diff
@@ -0,0 +1,111 @@
+
+Ensure you replace the `INSERT_MNEMONIC`, `INSERT_CONTRACT_ADDRESS` and `INSERT_ADDRESS_TO_CHECK` placeholders with actual values. Also, the contract ABI file (`Storage.json`) should be correctly referenced.
+
+## Conclusion
```
Maybe we can add a follow up to the REMIX IDE as well?

nhussein11 replied:
Sure, we can add REMIX IDE as a follow-up, but I see Remix more as a development environment. Ideally, we should link to some ethers.js tutorial that demonstrates how to implement a project using ethers.js. What do you think @CrackTheCode016 ?

In any case, those two options would be tasks to add later on.

CrackTheCode016 commented on develop/toolkit/api-libraries/smart-contracts/ether-js.md:
```diff
@@ -0,0 +1,111 @@
+
+## Introduction
+
+Ethers.js is a lightweight library that enables interaction with Ethereum Virtual Machine (EVM)-compatible blockchains through JavaScript. This article demonstrates how to use Ethers.js to interact and deploy smart contracts to the Asset Hub.
```
Maybe a word about how it works would be good, i.e., something like: 

"The `revive` compiler is used to compile Solidity smart contracts to PolkaVM bytecode, which allows it to be uploaded to Asset Hub. An Ethereum RPC faciliates interaction with existing Ethereum tools, such as Ethers.js or Metamask"

If this doesn't fit, feel free to ignore.

nhussein11 replied:
that's perfect! will added, thanks!

CrackTheCode016 commented on develop/toolkit/api-libraries/smart-contracts/ether-js.md:
````diff
@@ -0,0 +1,111 @@
+```
+
+!!! note
+    Replace `INSERT_RPC_URL`, `INSERT_CHAIN_ID`, and `INSERT_CHAIN_NAME` with the appropriate values. For example, to connect to Westend Asset Hub, you can use the following parameters:
````
```suggestion
    Replace `INSERT_RPC_URL`, `INSERT_CHAIN_ID`, and `INSERT_CHAIN_NAME` with the appropriate values. For example, to connect to Westend Asset Hub's Ethereum RPC instance, you can use the following parameters:
```

kianenigma commented on develop/toolkit/api-libraries/smart-contracts/ether-js.md:
```diff
@@ -0,0 +1,111 @@
+
+To deploy the compiled contract to the Asset Hub, you will need a wallet with a private key to sign the deployment transaction. You can use [subkey](/polkadot-protocol/basics/accounts/#using-subkey){target=\_blank} to manage your wallet.
+
+Here's the script to deploy the contract:
```
This is a fairly long script all in one go, can it be broken down into chunks?

kianenigma reviewed (commented):
The code snippets in this chapter are good examples to already be included in our strategy to be test-able right of the bat. We don't need to do it now, but cc-ing @UtkarshBhardwaj007 for discussion.

Note that my push to make things test-able will also force us all to use these tools e2e, which is good. 

Similar to the rust setup, we can bundle such codes into a single NPM project for each chapter. `npm run run` or similar should run a test script that would execute the flow e2e, and check the output.

Some notes about how to achieve this:

- We would need to rely on the console logs, capture them, and assert on them. For example, in this example, we should expect "Retrieved stored number: 42" to be in the logs
- We would need one seed account's private key to be stored in our gh as a secret with a lot of WND. This account can be used both for our manual testing, and in a CI to run this flow
- We should destroy the contract post test every time

Having written the above, it seems to me that this is non-trivial but achievable, so I won't necessarily suggest it now, but we should start thinking about it.

UtkarshBhardwaj007 commented on develop/toolkit/api-libraries/smart-contracts/ether-js.md:
```diff
@@ -0,0 +1,111 @@
+---
+title: Ethers.Js
```
NIT:
```diff
- title: Ethers.Js
+ title: Ethers.js
```

UtkarshBhardwaj007 commented:
> The code snippets in this chapter are good examples to already be included in our strategy to be test-able right of the bat. We don't need to do it now, but cc-ing @UtkarshBhardwaj007 for discussion.
> 
> Note that my push to make things test-able will also force us all to use these tools e2e, which is good.
> 
> Similar to the rust setup, we can bundle such codes into a single NPM project for each chapter. `npm run run` or similar should run a test script that would execute the flow e2e, and check the output.
> 
> Some notes about how to achieve this:
> 
> * We would need to rely on the console logs, capture them, and assert on them. For example, in this example, we should expect "Retrieved stored number: 42" to be in the logs
> * We would need one seed account's private key to be stored in our gh as a secret with a lot of WND. This account can be used both for our manual testing, and in a CI to run this flow
> * We should destroy the contract post test every time
> 
> Having written the above, it seems to me that this is non-trivial but achievable, so I won't necessarily suggest it now, but we should start thinking about it.

Agreed. I don't want to block this PR but a suggestion would be to write code that builds. Since we are creating tutorials, we could follow them to write code that actually builds and later on, as part of [this issue](https://github.com/polkadot-developers/polkadot-docs/issues/236) we can add tests to ensure that the code remains valid as testing (like you pointed out @kianenigma ) might be non-trivial. I am working on the `rust` code in the tutorials atm. I will pickup the `node` based code after that.

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,113 @@
+
+## Introduction
+
+The `revive` compiler is used to compile Solidity smart contracts to [`PolkaVM`](/develop/smart-contracts/evm/native-evm-contracts/#polkavm){target=\_blank} bytecode, which allows it to be uploaded to Asset Hub. An Ethereum RPC faciliates interaction with existing Ethereum tools, such as Ethers.js or Metamask
```
This first paragraph is weird because starts talking about `revive` compiler when the scope of the section is `Ethers`

Also, `An Ethereum RPC faciliates interaction with existing Ethereum tools, such as Ethers.js or Metamask` This should be maybe an introduction in the index page but not in here as we already are explaining what ethers can do

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,113 @@
+
+The `revive` compiler is used to compile Solidity smart contracts to [`PolkaVM`](/develop/smart-contracts/evm/native-evm-contracts/#polkavm){target=\_blank} bytecode, which allows it to be uploaded to Asset Hub. An Ethereum RPC faciliates interaction with existing Ethereum tools, such as Ethers.js or Metamask
+
+Ethers.js is a lightweight library that enables interaction with Ethereum Virtual Machine (EVM)-compatible blockchains through JavaScript. This article demonstrates how to use Ethers.js to interact and deploy smart contracts to the Asset Hub.
```
Add link to ethers docs

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,113 @@
+
+With the [`Provider`](https://docs.ethers.org/v6/api/providers/#Provider){target=\_blank} set up, you can start querying the blockchain. For instance, to fetch the latest block number:
+
+??? code "Fetch Last Block code"
```
I think this should be a plain code block without the collapsible. Also, add some inline comments to the code

nhussein11 replied:
I'm worried about this comment and [this one](https://github.com/polkadot-developers/polkadot-docs/pull/295#discussion_r1907613362).

I placed these portions of code in collapsibles because I was concerned about the page getting too long. @0xLucca 
 do you think it's worth having a long page but showing those snippets completely?

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,113 @@
+
+## Compile Contracts
+
+To deploy smart contracts to Asset Hub, you need to compile them into `polkavm` bytecode. Use the [`@parity/revive`](https://www.npmjs.com/package/@parity/revive){target=\_blank} library, which compiles Solidity code for use on Substrate-based chains.
```
```suggestion
To deploy smart contracts to Asset Hub, you must compile your Solidity code into `polkavm` bytecode. Use the [`@parity/revive`](https://www.npmjs.com/package/@parity/revive){target=\_blank} library to perform this step.
```

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
````diff
@@ -0,0 +1,113 @@
+
+Here's a sample Solidity contract (`Storage.sol`) to be compiled and deployed to the Asset Hub. This contract's functionality stores a number and permits users to update it with a new value.
+
+??? code "Storage.sol"
+
+    ```solidity
+    --8<-- 'code/develop/smart-contracts/evm-toolkit/ethers-js/Storage.sol'
+    ```
````
Probably no collapsible in here as the contract is part of the explanation

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
````diff
@@ -0,0 +1,113 @@
+--8<-- 'code/develop/smart-contracts/evm-toolkit/ethers-js/compile.js'
+```
+
+Note that the script above is tailored to the `Storage.sol` contract. It can be adjusted for other contracts by changing the file name or modifying the ABI and bytecode paths accordingly.
````
```suggestion
!!!note
    The script above is tailored to the `Storage.sol` contract. It can be adjusted for other contracts by changing the file name or modifying the ABI and bytecode paths accordingly.
```

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,113 @@
+
+## Contracts Deployment
+
+To deploy the compiled contract to the Asset Hub, you will need a wallet with a private key to sign the deployment transaction. You can use [subkey](/polkadot-protocol/basics/accounts/#using-subkey){target=\_blank} to manage your wallet.
```
```suggestion
To deploy the compiled contract to Asset Hub, you will need a wallet with a private key to sign the deployment transaction. You can use [subkey](/polkadot-protocol/basics/accounts/#using-subkey){target=\_blank} to manage your wallet.
```

0xLucca replied:
I don't think subkey is the best example here, as you could use any other tool that's being used in evm development

0xLucca commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
````diff
@@ -0,0 +1,113 @@
+Here's the script to deploy the contract:
+
+```js
+--8<-- 'code/develop/smart-contracts/evm-toolkit/ethers-js/deploy.js'
````
This script is really long, so you can use multiple code snippets explaining each part and then show the full code snippet below

0xLucca commented on .snippets/code/develop/smart-contracts/evm-toolkit/ethers-js/checkStorage.js:
```diff
@@ -0,0 +1,84 @@
+const mnemonic = 'INSERT_MNEMONIC';
+const contractName = 'Storage';
+const contractAddress = 'INSERT_CONTRACT_ADDRESS';
+const checkAddress = 'INSERT_ADDRESS_TO_CHECK';
```
What's this variable for?

nhussein11 replied:
I forgot to remove this unused variable

0xLucca reviewed (approved):
LGTM!

dawnkelly09 commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,146 @@
+
+After executing the script, the Solidity contract will be compiled into the required `polkavm` bytecode format. The ABI and bytecode will be saved into files with `.json` and `.polkavm` extensions, respectively. You can now proceed with deploying the contract to Asset Hub network, as outlined in the next section.
+
+## Contracts Deployment
```
```suggestion
## Deploy the Compiled Contract
```

dawnkelly09 commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,146 @@
+
+The deployment script can be broken down into key components:
+
+1. First, set up the required imports and utilities:
```
Where does this code go? Is part of an existing file or do I need to create a new one? 

Please add titles to the code fences for all of these snippets so the user knows which file you're working in.

dawnkelly09 commented on develop/toolkit/api-libraries/smart-contracts/ethers-js.md:
```diff
@@ -0,0 +1,146 @@
+
+Ensure you replace the `INSERT_MNEMONIC`, `INSERT_CONTRACT_ADDRESS` and `INSERT_ADDRESS_TO_CHECK` placeholders with actual values. Also, the contract ABI file (`Storage.json`) should be correctly referenced.
+
+## Conclusion
```
Let's trade this out for a ## Where to Go Next for consistency with our other pages. See example page: /develop/toolkit/api-libraries/py-substrate-interface/#where-to-go-next

dawnkelly09 reviewed (changes requested):
Needs a little love to give a more polished feel. Thank you!

nhussein11 commented:
@dawnkelly09 I've addressed your comments and I also aligned this branch with the (temporal) architecture present in https://github.com/polkadot-developers/polkadot-docs/tree/staging-ah-smart-contracts. This should be ready for final review now

nhussein11 commented:
> > @dawnkelly09 I've addressed your comments and I also aligned this branch with the (temporal) architecture present in https://github.com/polkadot-developers/polkadot-docs/tree/staging-ah-smart-contracts. This should be ready for final review now
> 
> When I spun this up locally, the ethers.js page only renders the title and the feedback stuff. It's weird. ![ethers1](https://private-user-images.githubusercontent.com/83190195/404974054-497f1e44-5787-4cba-a228-0dfd1928a3df.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczOTc3MzYsIm5iZiI6MTczNzM5NzQzNiwicGF0aCI6Ii84MzE5MDE5NS80MDQ5NzQwNTQtNDk3ZjFlNDQtNTc4Ny00Y2JhLWEyMjgtMGRmZDE5MjhhM2RmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIwVDE4MjM1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ4ZjllMDZjY2Q3MzA4N2ZlMTRhM2VlN2RlOThiODM4YTVhODhmYTJhNmZjNjVjMmQ2NTQ2NDNhZGVjODM5OWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9GVjQb5fnauH91HgrWt2TxHkSxUhh46yXJTx-ujMcZ8)

Should be fixed in [103fa77](https://github.com/polkadot-developers/polkadot-docs/pull/295/commits/103fa77f3b0278275e8b02666a8ebb2c0d2386ed)

dawnkelly09 commented on develop/smart-contracts/evm-toolkit/libraries/ethers-js.md:
```diff
@@ -0,0 +1,154 @@
+Now that you have the foundational knowledge to use Ethers.js with Asset Hub, you can:
+
+- Dive into Ethers.js utilities - discover additional Ethers.js features, such as wallet management, signing messages, etc
+- Implement batch transactions - use Ethers.js to execute batch transactions for efficient multi-step contract interactions
```
```suggestion
- **Implement batch transactions** - use Ethers.js to execute batch transactions for efficient multi-step contract interactions
```

dawnkelly09 commented on develop/smart-contracts/evm-toolkit/libraries/ethers-js.md:
```diff
@@ -0,0 +1,154 @@
+
+- Dive into Ethers.js utilities - discover additional Ethers.js features, such as wallet management, signing messages, etc
+- Implement batch transactions - use Ethers.js to execute batch transactions for efficient multi-step contract interactions
+- Build scalable applications - combine Ethers.js with frameworks like [`Next.js`](https://nextjs.org/docs){target=\_blank} or [`Node.js`](https://nodejs.org/en){target=\_blank} to create full-stack decentralized applications (dApps)
```
```suggestion
- **Build scalable applications** - combine Ethers.js with frameworks like [`Next.js`](https://nextjs.org/docs){target=\_blank} or [`Node.js`](https://nodejs.org/en){target=\_blank} to create full-stack decentralized applications (dApps)
```

nhussein11 commented on develop/smart-contracts/evm-toolkit/index.md:
```diff
@@ -1,9 +1,11 @@
 
-TODO: complete this index page
\ No newline at end of file
+TODO: complete this index page
```
Mmm idk, to be honest. Since we are merging this into a branch that has a temporary Information Architecture (see this comment https://github.com/polkadot-developers/polkadot-docs/pull/300#issuecomment-2602770580), many things will probably change, particularly regarding indexes and so on. 

I was considering making a separate PR when we have the final IA to update not only the website structure but also the index pages. WDYT, @dawnkelly09?

nhussein11 commented on develop/smart-contracts/evm-toolkit/libraries/ethers-js.md:
```diff
@@ -0,0 +1,154 @@
+
+## Compile Contracts
+
+The `revive` compiler transforms Solidity smart contracts into [`PolkaVM`](/develop/smart-contracts/evm/native-evm-contracts/#polkavm){target=\_blank} bytecode for deployment on Asset Hub. Revive's Ethereum RPC interface allows you to use familiar tools like Ethers.js and MetaMask to interact with contracts.
```
you are right, I missed this one

dawnkelly09 reviewed (approved):
Approved for staging but needs work on index page(s) before it can be deployed publicly.

eshaben commented on develop/smart-contracts/evm-toolkit/libraries/ethers-js.md:
```diff
@@ -0,0 +1,154 @@
+---
+title: Ethers.js
+description: Learn how to interact with the Asset Hub chain using Ethers.js, compiling and deploying Solidity contracts, and interacting with deployed smart contracts.
```
```suggestion
title: Deploy Contracts to Asset Hub with Ethers.js
description: Learn how to interact with the Asset Hub chain using Ethers.js, from compiling and deploying Solidity contracts to interacting with deployed smart contracts.
```

We need to be more descriptive for SEO purposes. Just Ethers.js will get lost in all the other results specific to ethers

eshaben commented on develop/smart-contracts/.pages:
```diff
@@ -5,4 +5,4 @@ nav:
   - 'Parachain Contracts': parachain-contracts.md
-  - 'EVM Toolkit': evm-toolkit
\ No newline at end of file
+  - 'EVM Toolkit': evm-toolkit
```
```suggestion
  - evm-toolkit
```

We don't typically provide names at this level for directories

eshaben reviewed (commented):
One thing to consider is putting the variables at the top of the files, so users when they first look at the code snippets, they instantly see that they need to update those
